src/ORM/Rule/ExistsIn.php

```diff
@@ -98,6 +109,11 @@ public function __invoke(EntityInterface $entity, array $options)
             return true;
         }
 
+        if ($this->_options['allowSqlNulls'] === true
+            && $this->_checkPartialSchemaNulls($entity, $source) === true
+        ) {
+            return true;
```

Comment:
Before I (try to) swap the fixtures, take a look at the test, fixture and example use case and it should be (almost) obvious why.

For an ExistsIn() constraint where the referenced composite key may be partially null if just one of the keys that MAY BE null IS NULL then the whole rule should return true. If you need to be more specific you want to:
- Setup additional ExistsIn() constraints on smaller parts (like only site_id),...
- ... or set all fields to be NOT NULL.

So yes, the intended behavior is such that the top rule will pass because it cannot exist anyway because parent_id is null. If you don't want to check if for instance a Category, given a Site exists in a ParentCategory (given the same Site) but only want to check if a Category references a correct and existing Site then that's what you do:

```php
        $rules->add($rules->existsIn(['site_id'], 'Sites')); // Does this SiteCategory reference an existing Site.
        $rules->add($rules->existsIn(['parent_id', 'site_id'], 'ParentNodes', ['allowSqlNulls' => true])); // Does this SiteCategory reference and existing Parent SiteCategory or none at all (parent_id = null)
```